### DISCUSSION TITLE
Service Mode Button for Nixie Standalone

### DISCUSSION THREAD

**Discussion by rock-crusher** [tagyoureit/nodejs-poolController#508] [Nixie]
Just wondering if this has ever been implemented, and if not is there any way possible to do something similar for the Nixie standalone setup? 
It sure would be nice to have a physical button inside or on the outside of the controller enclosure that would make it easy for someone to put the system into service mode.
Something simple that would give a person full control of all the equipment manually at the pad.

**Comment by tagyoureit** [comment]
We don't really have a service mode [button].  I typically just stop the process. You could wire up the button to stop njsPC and then restart it...

**Reply by orrious** [reply]
I look every so often for display / control for my panel, but have yet to find something I'm willing to invest in.  Has anyone else created a display for their pad controls?

**Reply by rock-crusher** [reply]
> We don't really have a service mode [button]. I typically just stop the process. You could wire up the button to stop njsPC and then restart it...

Example for: "I typically just stop the process"
Example for: "wire up the button to stop njsPC and then restart it..."

Sorry tagyoureit, but I am very illiterate when it comes to this.
Thank You

**Reply by tagyoureit** [reply]
> Example for: "I typically just stop the process"

How are you starting the process?  I use pm2 and `pm2 stop njsPC` does the job.

> Example for: "wire up the button to stop njsPC and then restart it..."

There are lots of examples out there for wiring a button to shutdown the Pi like [this one](https://www.makeuseof.com/tag/add-reset-switch-raspberry-pi/).  You could start with that but modify the script to stop/start njsPC instead of shutting down the whole Pi.

**Reply by rock-crusher** [reply]
Hi again tagyoureit, although the "pm2 stop njsPC" works just fine for shutting things down, it does not allow for manual control of the pump at its keypad.

Also, a complete shutdown of the pi only seems to turn off the pump and leaves its key pad disabled. During this shutdown the relays 
(Sequent 8 hat) that were already turned on prior to shut down stay on after the pi shuts down, this seems weird. 

Unfortunately, this idea wont work for a service mode button, unless after shutdown the pump keypad becomes enabled for use.

**Reply by rstrouse** [reply]
Eventually, the pump will release after a period of time.  It requires a heartbeat request from njsPC to remain under control.  When I get some time I look to see if there is a message to expedite this release and add it to the shutdown.

**Reply by tagyoureit** [reply]
We are actually doing the proper shutdown in the code.  This is not working with PM2 because PM2 has a [default shutdown timeout of 3s](https://pm2.keymetrics.io/docs/usage/signals-clean-restart/).  It is killing the njsPC process before it has a chance to shut down the pumps properly.  

Grab the latest [ecosystem.config.js](https://github.com/tagyoureit/nodejs-poolController/wiki/Automatically-start-at-boot---PM2-&-Systemd).  It has a new `kill_timeout` value of 15s which is more then enough to shutdown njsPC properly.  Mine takes about 7-8s.  

To put this file into effect, run Steps 4 & 5 on the link for the file.

**Reply by rstrouse** [reply]
Wooomp there it is

**Reply by rock-crusher** [reply]
Hi again tagyoureit / rstrouse, thanks again for all the help you have provided me along this journey.

I went ahead and put in the new script and it works just fine.

Problem for me now is how to implement a physical push button to start the process “pm2 stop njsPC” and of course then to implement the “pm2 start njsPC” using a second physical push button.

tagyoureit, I checked out the link you provided, but have no clue how to convert this tutorial into what I would like to achieve above.

As you know, I haven’t the knowledge to implement this on my own and sure would appreciate any help I can get.

I am a little surprised though that a service button has never been incorporated into the Nixie, or even as an option during the set-up procedure.

Being able to have manual control of the equipment out at the pad with a quick press of a button seems to be a great option and would be handy for anyone needing manual control such as a pool service technician even if that person is you.   

I would think others might find this option useful as well… but that’s just me.

Thanks again guys for the great work!

**Comment by rstrouse** [comment]
Service and Timeout modes are probably not a bad addition to Nixie.  I'll take a look this afternoon and see what it might take to make it happen.

**Comment by rstrouse** [comment]
Ok so now there is a Service and Timeout mode for Nixie controllers.   If you pull njsPC, dashPanel, and REM you will find the option by clicking on the status area in the dashPanel header.
![image](https://user-images.githubusercontent.com/47839015/176563108-75d4618d-b061-4004-9fc3-58a56cc2b995.png)

To automate this through REM a new feed was created that allows you to hook it up to a button.  I hooked it up to a GPIO momentary button.  The feed below will toggle between Service Timeout and Auto with a timer expiring 1 hour after the timeout was started.  This way the panel can be set to service mode for a period of time and then automatically go back to auto.
![image](https://user-images.githubusercontent.com/47839015/176563419-420c5b6c-3841-4301-a658-c9aef4b4f71d.png)

**Reply by rock-crusher** [reply]
Hell yes rstrouse!
Can’t wait to try this out!
I think it’s a great addition to Nixie.
Thanks again for all your hard work and effort!

**Reply by rock-crusher** [reply]
Hi again rstrous, 
I'm having some difficulty with the service mode, I did the pull on njsPC, dashPanel, and REM along with a hard refresh.

The issues are as follows:
1. I am not seeing the new feed that was created “Edit GPIO Device Feed”

Maybe I didn't pull REM correctly, or?

2. Everything in dashpanel shows up, so I did a trial run with it. Everything shutdown as expected, but when timed out, 
or when clicking the “Auto” button to return to auto mode it did not resume the existing running schedule, or anything
that was running during the time of putting it into service mode/timeout.

I did a reboot of the pi and a few seconds later the running schedule resumed, so I decided to do another trial run, but same scenario, only way it resumes is by rebooting the pi.

Could this be a result of my REM  issue?

**Reply by rstrouse** [reply]
It will not resume the existing schedule if the schedule is past the start date.  I thought about drawing outside of the normal functions we see on OCPs to have it restart scheduled items but not manually stopped items.

I am guessing for REM you are not selecting the right options for the feeds.  Are you attaching this to a GPIO pin or something else?

**Reply by rock-crusher** [reply]
That was another thing I wanted to ask you.
As you know, I would like to have a physical momentary button to automate this, but don't know how to go about it:

1. where to connect the button lead wires

2. what needs to be done to make the button work

I’m sorry rstrouse, this is probably stupid easy for all the other guys here, but for me… 
well, NOT!

I cant even find this window:

![176563419-420c5b6c-3841-4301-a658-c9aef4b4f71d](https://user-images.githubusercontent.com/94593954/176780483-89641423-bc16-4446-ad7b-a51773402be2.png)

**Reply by rstrouse** [reply]
The first thing we have to do is figure out where you want to hook up your button.  Do you have a MEGA-BAS?  If you do it becomes really easy.  If you do not then it is still easy it just requires a few more steps.  Either way wire them just like you would wire a flow switch.  Replace the flow switch with a momentary button and you are off and running.

https://github.com/tagyoureit/nodejs-poolController/wiki/DIY-REM-Chem-Controller#wire-the-flow-switch

You will find a Feeds tab on all devices in REM.  So once you have the device selected (GPIO pin, BAS... etc) the tab interface will include a Feeds tab.  On the upper right hand corner of the Device Value Feeds list you will see a + symbol to add a new feed for the device.  If you are using a GPIO pin, first set the pin direction to Input, then hook up the button as described above.  If you are using a MEGA-BAS configure one of the 0-10v inputs on the left hand side of the board as a Dry Contact by setting the appropriate jumper and then configure it in REM under the Inputs tab.

Either way the only difference between the two configurations is the send value.

**Reply by rock-crusher** [reply]
Hi again rstrouse,

Just got around to trying out the external push button control, seems to be working as stated, and yes, I am using the MEGA-BAS.
Thank you for that!

I played around adding a button for each of the modes and it worked great. 

I did notice that the service mode option allows you to also select timing, not sure why though because no matter what its set at it will run 24/7, I am assuming this is correct, at least until its put back into auto mode/timeout mode?

Another question I have is: would it be relatively easy (for me) to wire up 3 leds so they will be in sync with dashpanel display (in real time) when in auto mode, service mode, or timeout mode. A total of 3 leds indicating which mode is in play in real time.

This would be helpful for a service person out at the pad since there would be no display out there for them to look at.

I would like to use [this type led pushbutton](https://www.amazon.com/dp/B097T7VSVT/ref=sspa_dk_detail_5?pd_rd_i=B097T7VSVT&pd_rd_w=ItrjA&content-id=amzn1.sym.3be1c5b9-5b41-4830-a902-fa8556c19eb5&pf_rd_p=3be1c5b9-5b41-4830-a902-fa8556c19eb5&pf_rd_r=CV6AQE054XSN43QWD717&pd_rd_wg=NZrSu&pd_rd_r=1b335664-2b91-477d-b494-0fddde86d039&s=pc&spLa=ZW5jcnlwdGVkUXVhbGlmaWVyPUEyOVAxOUFBQldZVkRDJmVuY3J5cHRlZElkPUEwMzc2NDU5M1ZCV0VKUExKSzNIUyZlbmNyeXB0ZWRBZElkPUEwNDA2NzU3TDRFMlFLMDBTSVkwJndpZGdldE5hbWU9c3BfZGV0YWlsJmFjdGlvbj1jbGlja1JlZGlyZWN0JmRvTm90TG9nQ2xpY2s9dHJ1ZQ&th=1), or similar if it would work for each mode out at the pad:

**Reply by rock-crusher** [reply]
…or would it be better to use an led alongside the pushbutton?

… or maybe there is a better way to achieve this altogether?

Thanks again rstrouse

**Reply by rstrouse** [reply]
If you are up to date it will always be in sync.  The two modes where you can use the timeout are Timeout and Toggle.  If you use toggle and the time is 0 then it will toggle between service and auto on each button push.  If you put a time in there it will return to auto mode after the specified time but if you want to return to auto before the timeout you can hit the button again.

**Reply by rstrouse** [reply]
You need to find a button or led that can be driven by led then simply create an output that is triggered by the same njsPC binding.  So you would use a trigger instead of a feed.

**Reply by rstrouse** [reply]
Here is a button that has a 3v LED in it instead of the 12v variety.  Ensure you install a resistor in the wiring to make sure you don't cook the LED.  It is in the instructions for it.
https://www.adafruit.com/product/1477?gclid=EAIaIQobChMI1cDervbf-AIVyB6tBh1vbQfGEAQYAiABEgLGAfD_BwE

**Reply by rock-crusher** [reply]
Hi rstrouse,

I didn’t expect any feedback today, but sure appreciate it, Thanks! 

Yes, that is what I would like to use for this project, 3 simple pushbuttons with indicating led for the service, timeout, and auto modes out at the pad.  

Most service personnel are familiar with those modes and having this available would make it very quick & easy for them to implement and do their work.

For those that use a service technician for weekly maintenance, filter teardowns, repairs etc. having this set-up out at the pad would be a must... don't you think?

Happy 4th rstrouse and to anyone else reading this here today!

**Reply by rstrouse** [reply]
Oh I was suggesting a single button that toggles between Auto and Timeout or Auto and Service.  Such a thing could reflect the current mode regardless of where the mode was started from using the LED on the button and a trigger on the LED pin.  

So you would press the button once and it would change the panel to service mode and illuminate the button.  Then press it again and it would return to auto and turn the LED off.   If you engaged timeout or service from any other place like dashPanel it would illuminate the button but you could cancel it by pressing the button on the pad.

You could still use 3 buttons or 2 buttons even.  If you press the timeout button it will toggle to timeout or if you press the service button it will toggle to service.  The reason there is such a thing as timeout is because service peeps can put chems in the pool and want to keep the pump running for a period of time but won't hang around or return.  So the timeout is triggered and there is no danger that the pool will not return to its regularly scheduled program.  

The tough part with that is resuming schedules.  I wouldn't resume manual triggered circuits but I am on the fence with resuming the schedules.  While in either service mode it disables others from starting circuits on the pool.  But if it suddenly pops out of service mode and returns to the regularly scheduled program it could be unfortunate.   Maybe that is why Pentair doesn't start anything until you re-engage the circuits or the start time is hit again.

I have seen peeps talk about their process out on TFP where they put the pump into service mode which stops it then promptly disassemble the filter to clean it.   If the timeout simply expires no danger.  If it engages the pump it could be bad.

**Reply by rock-crusher** [reply]
I understand your hesitance on resuming the schedules.  Maybe that could be an option one could choose in the set-up?

The 2 button set-up you mention sounds really good! 

Is this something you can walk me through?

**Reply by rstrouse** [reply]
Sure use the dry contact wiring on the MEGA-BAS for both buttons and set up a feed as described below for each of them.

The key is to set up one button with Mode set to Toggle and the desired timeout time and the other button as Mode set to Toggle with a Timeout of 0.  This will toggle the state between Auto and Timeout with the one that has a Timeout set and Auto and Service for the one without.  
![image](https://user-images.githubusercontent.com/47839015/177418302-cd4aede8-a359-494b-8b89-6a0fd5d53c68.png)

Once you get the buttons doing their thing I will walk you through setting up an indicator LED for the modes.

**Reply by rock-crusher** [reply]
Awesome rstrouse thank you, works just as you said!

Now for some detailed instruction on the leds and should be good to go… I hope.

Pictures really help me understand what I’m doing, thanks for that.

**Reply by rstrouse** [reply]
So are you planning two leds or one?  Which LEDs are you using?

**Reply by rock-crusher** [reply]
Well, 2 for sure, but is it possible to use 3, the third to indicate auto mode?

I don’t have the pushbuttons with built-in led, but do have a bunch of individual 5mm red leds to work with along with several momentary push buttons.

So, at this point I will use the pushbuttons I have with the red indicating led alongside it.

**Reply by rstrouse** [reply]
Yes you can use up to 3 to indicate which mode the panel is in.

I added the triggers for panelMode so pull REM.  Then you need to choose as many GPIO pins that you want to use for the LEDs.  You can drive all 3 of these LEDs off the GPIO pins on the Pi and any unused GPIO can be used.  

Use 330-470 ohm resistors if you are using RED leds.  For any other color use 470ohm to 1kohm resistors for any other color.  This will drive the LEDs so that they are bright but will not overdrive the LED or the pin.  Make sure you hook the longer wire of the LED to the anode (+).  This will be the wire that runs from the GPIO pin through the resistor.  The shorter wire on the LED should hook to the ground.
![image](https://user-images.githubusercontent.com/47839015/177584991-f67bc518-537a-492c-bcc3-e7461cf24863.png)

After you have hooked up the LEDs to the pins go into REM and configure a Trigger this time that will change the LED state when the state changes.  You will need 2 triggers per pin.  One trigger will turn it on and the other trigger will turn it off.  

Here is an example for a trigger that turns on the LED when auto is detected.  The key thing here is that the Desired State is set to On and the mode.name filter is checked with the value == auto.  There are three possible values here auto, service, or timeout all in lower case.
![image](https://user-images.githubusercontent.com/47839015/177586818-88d41c41-b89f-46f0-a286-9cf396f9b68f.png)

Here is an example for a trigger that turns off the LED when Nixie is not in auto mode.  The desired state is Off and the mode.name is set to not equal to auto.
![image](https://user-images.githubusercontent.com/47839015/177587510-615a2cef-9de6-44dd-b7b3-76bd781129bd.png)

When you are done setting up the auto LED like below, set up the two other pins but use service for one and timeout for the other.
![image](https://user-images.githubusercontent.com/47839015/177587859-14b54985-8d7a-4063-bdce-49dac7a1095a.png)

**Reply by rock-crusher** [reply]
Hi rstrouse, well, everything works except the service led.

I double checked everything several times, and then redid it all and still the service led is the only one that does not work.

I copied your work just as the pictures show, and then added the service and timeout led.

The auto led and the timeout work perfect from both the physical pushbutton and from dashpanel.

Just the service led no work?

**Reply by rstrouse** [reply]
Pull njsPC, REM, and dashPanel.  I added some code to cancel delays when entering into Service Mode.  Let's make sure you are running on the most recent software.

Which pin number did you use for service mode?

**Reply by rock-crusher** [reply]
I did a pull on njsPC, REM, and dashPanel

I'm using pin 24 for service mode, but still doesn't work?

![Screenshot from 2022-07-06 17-29-37](https://user-images.githubusercontent.com/94593954/177665688-a16293da-3dc8-4a2c-b3f1-a909f28d5e4f.png)
![Screenshot from 2022-07-06 17-30-42](https://user-images.githubusercontent.com/94593954/177665709-807acecf-24bd-4585-ad88-b526d9183dcf.png)
![Screenshot from 2022-07-06 17-28-29](https://user-images.githubusercontent.com/94593954/177665728-c712599b-8dfd-4813-8941-d5398471ef07.png)

**Reply by rstrouse** [reply]
You typed service with a space after it.  Remove the space and it should work.

**Reply by rock-crusher** [reply]
Oh my gosh! Good eye!

Didn’t even notice, and even if I did, I wouldn’t have thought an empty space at the end would make any difference.  I did it on both on and off… lol  WTF, have no Idea how I managed that and I redid it several times to boot!

It does work great now, THANK YOU!!

Question:  How does one check voltage on an LED with no specs? The LED I have on hand that I’m using are random 5mm red ones and with the 330 ohm resistors they seem very dim.

**Reply by rstrouse** [reply]
Forward operating voltage should be between 1.8 and 2.2v @20mA.  You can get a bit brighter with a 220ohm resistor.  If that isn’t bright enough get the super bright variety.

**Reply by rock-crusher** [reply]
Ok thanks, yeah these are very very old. I tried the 220 ohm and still dim.

Probably not wise to try without a resistor at all?

I’ll get some new ones and give it another go.

Anyway, thank you again very much for helping me out!

**Reply by rstrouse** [reply]
Do not do that you will overdrive the pin!

**Reply by rock-crusher** [reply]
Just an update, I replaced the old LEDS and all are nice and bright now even when using 1,000 ohm resistors.

Just for the experience, I decided to try the alternate connection for the pushbuttons “Method 2” using GPIO pins. This method (in my opinion) the pushbuttons are far more responsive in activating the different modes (less lag time). The press of the button just seems to be more accurate & crisp as opposed to when connected to the MEGA-BAS. 

Needless to say, I’ll choose “Method 2” for my project.

Again, thank you rstrouse for all the help you’ve provided me!

**Reply by rstrouse** [reply]
Yep either one works.  The code for the gpio is not bound to the poll so it will be very responsive.

**Reply by rstrouse** [reply]
When you get it all mounted up shoot some pics in the My Pool section.

**Reply by michaelmedellin** [reply]
@rstrouse this is likely a very silly question, but how do I leverage the built in button on the SMS 8-Relay Hat on GPIO 26 to toggle service mode at the pad? I have the current configuration attached, but I seem to be missing a step because this isn’t toggling modes.

![84496CC4-81E8-48F0-A836-F6B91DA42C56](https://user-images.githubusercontent.com/13091665/213286403-f2557949-c351-45ad-abb3-d87834b6f0da.jpeg)

**Reply by rstrouse** [reply]
Does the pin state change when you press the button on the relay card?

**Reply by michaelmedellin** [reply]
It does not. 

I do know I can monitor the pin state with a python script. So the button is working. Just unsure on the integration with REM. Cheers.

**Reply by rstrouse** [reply]
Restart the pi after you have this configured.  It is possible that the pin is occupied by another process and the interrupt is not being triggered.

**Reply by michaelmedellin** [reply]
Ah, I think my issue is that I’m running REM as a non-privileged container so need to pass in the appropriate devices/permissions to access GPIO. Will update if I find a solution.

**Reply by michaelmedellin** [reply]
Yep, that was the issue. For the GPIO library used with REM (onoff), I needed to mount the following volumes when running an unprivileged container:
- /sys/class/gpio 
- /sys/devices/platform/soc/fe200000.gpio/gpiochip0/gpio/gpio26

Now it’s working. Thanks for the help @rstrouse.

**Comment by rstrouse** [comment]
> @orrious said: I look every so often for display / control for my panel, but have yet to find something I'm willing to invest in. Has anyone else created a display for their pad controls?

Here is my control panel at the pad.  I made a dead panel for my controller out of 3/8" ABS and mounted it there but don't really use it all that often since I can simply use an ipad.
![image](https://user-images.githubusercontent.com/47839015/176567789-e8b45f94-4443-406e-9cbd-d5b6b67ff02e.png)

**Reply by rock-crusher** [reply]
I like this idea, very clean install!

**Comment by mhmatttu** [comment]
I have setup nixie and added physical buttons using the Adafruit button recommended in this post.  I have confirmed that the buttons are working through the MEGA-BAS inputs tab and the LEDs work when the respective circuit is activated.  I setup one button to be service mode using the above instructions and works great (except for not being very fast to respond).  However, I have not yet been able to get other circuits to work properly with the buttons.  Below are the feeds for my pool circuit and light circuit.  When pressing the button, the respective circuit comes on but then quickly cycles off.  I got the circuit ids from /state/all and assume they are correct since the circuit does activate.  What am I doing wrong? @rock-crusher did you setup other buttons besides Modes?

Pool Body
![image](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/03260526-792b-4b4a-80e1-a29f3f419029)


Light Circuit
![image](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/a7ac9825-1e49-4162-8aa3-19a9f6ebd021)

**Reply by mhmatttu** [reply]
Another update.  I attempted the GPIO method mentioned above wiring like a flow switch with resistors.  Below are the settings.  I get the same result of the circuit coming on with the button press but turns off as soon as I release the momentary button.  If I hold the button down, the circuit stays on so must have something to do with the momentary button part of it.  This behavior does not happen with the service mode button I setup using above instructions. 

I currently have this all on my desk and just waiting on this last piece to install at my friends house.  Appreciate any thoughts/help on this!  Thanks.

![image](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/3f2c8a90-1347-4551-a92b-c955ca610561)

**Reply by mhmatttu** [reply]
@tagyoureit @rstrouse Good afternoon gentlemen.  Curious if you have any thoughts on what I may be doing wrong with my setup? It almost seems like it needs a toggle mode similar to the panelMode socket event offers.  Appreciate the help!

**Reply by rstrouse** [reply]
If you are using a momentary button the GPIO will go low as soon as you release it and therefore turn it off.  We could add a toggle to the input that would cover using a momentary button.  First push to turn it on and second to turn it off.

**Reply by mhmatttu** [reply]
That would be awesome!

**Reply by rstrouse** [reply]
Pull njsPC and REM the momentary pushbutton code has been implemented as a Toggle property.

**Reply by mhmatttu** [reply]
Works perfectly!!  Thank you very much!

**Comment by mhmatttu** [comment]
Couple pics

![IMG_2874](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/cc83164d-ab6d-46bc-b87e-b087b5bf6256)

![IMG_2896](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/f994302b-c8ac-4d01-b4e3-d835bbef68a1)

**Reply by rstrouse** [reply]
Awesome!  That looks like even a caveman could operate it at the pad.

**Comment by brentk7** [comment]
Hi @rstrouse , I am trying to implement mode change using pushbuttons connected to Digital inputs on a Sequent 4rel4in v4 card.
As soon as I add the feeds as described above - I get the following errors in my logs despite the fact that on the Inputs tab of the web interface - the green indication comes on when the button is pressed (and the relays work fine). Any idea what is wrong here?
```
0|REM        | [18/06/2024, 17:23:37] error: Error getting Sequent 4rel4in v4 digital input value for inDigital1
0|REM        | [18/06/2024, 17:23:37] error: Error getting Sequent 4rel4in v4 digital input value for inDigital2.value
```
![image](https://github.com/tagyoureit/nodejs-poolController/assets/25597360/fce30827-708c-427c-8454-6d9ce55e5008)

**Reply by brentk7** [reply]
Also getting another problem when I push SAVE after enabling GPIO26 to try use the pushbutton on the relay card:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/25597360/1d54dee6-342d-4f3d-bf7d-ff138bd750d9)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/25597360/28f5a76c-7536-4b5d-8660-6b53a7e604e2)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/25597360/fab656dd-cb80-4ccd-96cd-fafa31650dc2)

**Reply by rstrouse** [reply]
If you upgraded to the latest os for the pi you need to select the proper os selection on the general tab.  The pi foundation really screwed up on the gpio mapping.

**Reply by brentk7** [reply]
Thanks for the reply. This is a fresh install of both OS and njsPC/REM.
Do you mean the general tab of REM? I only have the option for "Raspberry Pi" (or Pi 4B) there (I am using Pi 3 Model B Rev 1.2 with Debian GNU/Linux 12 (bookworm)). 
Should I choose the Pi4B Bookworm even though I am running on a 3B?

![image](https://github.com/tagyoureit/nodejs-poolController/assets/25597360/8b2900ae-d197-46b6-a387-884c3066f15f)

**Reply by brentk7** [reply]
Changing the Board selection to "Raspberry Pi 4B Bookworm OS" does get rid of the error for GPIO26, but I still have the error 
 
Error getting Sequent 4rel4in v4 digital input value for inDigital1.value

Any suggestions for this one?

**Reply by mhmatttu** [reply]
Hi @brentk7.  In looking at your screenshot above, it looks like you chose “inDigital1” vs “inDigital1.value” for your first feed.  I made this mistake initially when setting up temp sensors and changing to the .value fixed it for me.

**Reply by brentk7** [reply]
Hi @mhmatttu thanks for that, unfortunately that is not the fix for mine. I had tried both and they both give the same error (in fact if you look at my screenshot and error logs - I am looking at inDigital1 and inDigital2.value and both give the same error).
I have also tried with only inDigital1.value, but get the same error

**Reply by brentk7** [reply]
Thanks @rstrouse [Fix IO array references](https://github.com/rstrouse/relayEquipmentManager/commit/7dff1adaf26abc28f014c23c3d7a59e30fe64aae)  has fixed this for me. 
Just a note that the sequent card seems very slow to register the input - I have to hold the button in for at least a second. I assume that this is because the button is only polled every second or so.

**Comment by dkossman** [comment]
Hey guys, I have an issue related to this discussion - i've actually logged a bug but am not sure whether it's a config issue peculiar to my system.  I have the service mode button implemented, and it toggles the system in and out of service mode as expected.  however, when it returns to auto mode, the current schedule is not resumed.  is this just me?  I see the same behavior if i toggle back to auto mode from Dashpanel - the system does not re-trigger the currently active schedule.  any suggestions would be appreciated.

see https://github.com/tagyoureit/nodejs-poolController/issues/1032

**Reply by mhmatttu** [reply]
Good morning.  Not sure if you saw this post above (https://github.com/tagyoureit/nodejs-poolController/discussions/508#discussioncomment-3059431) and if it relates to your issue.  I believe I saw it explained in more detail in another thread as well.

**Reply by dkossman** [reply]
Ah - thanks.  I hadn’t expanded that discussion and now see the rationale explained by @rstrouse for not resuming schedules.  However in my case the button does not trigger a timeout, it just toggles service mode.  It would be great if there could be a config option to govern the resume behavior… I’ll update the issue I logged.  If a switch was used instead of a toggle button I think that might alleviate the concern about unexpected resumption of equipment.

**Reply by mhmatttu** [reply]
I dont have the timeout feature enabled but it is there.  @rstrouse mentions it in this [post ](https://github.com/tagyoureit/nodejs-poolController/discussions/508#discussioncomment-3052151).  I believe you can setup to toggle between service and timeout or could have 2 buttons to control separately.  I think @rock-crusher used 1 button with 3 LEDs.  The 1 button toggled between Auto, Service and Timeout and the LEDs indicated which mode it was in.  The pool panel I configured for a friend uses the buttons mentioned here but my pool is a Jandy Aqualink panel with njsPC integration.  The Aqualink also has a single button for Auto, Service, Timeout with an LED indicating which mode is active.

**Reply by dkossman** [reply]
I was thinking that if i use  an on/off switch rather than a button, and set the feed mode for the GPIO to "service" rather than "toggle", it would be reasonably safe to allow the system to resume schedules when service mode is turned off (via the physical switch).  However need to confirm with @rstrouse that
- he would be willing to update NJSPC with a config option to resume schedules when Auto mode is resumed
- the system would read the switch state on startup so go into service mode if it is on

I've updated the issue i logged as a feature request...

<img width="1162" alt="Screenshot 2024-10-03 at 1 08 35 PM" src="https://github.com/user-attachments/assets/c19ec5da-a10c-41db-87e5-31b049828bbf">
